analysis_driver/clarity/__init__.py

```diff
@@ -96,41 +98,63 @@ def get_species_information_from_ncbi(species):
             if match:
                 common_name = match.group(1)
             all_species_names.append((taxid, scientific_name, common_name))
-    if len(all_species_names) > 1:
-        app_logger.error('More than one taxon corresponding to %s', species)
-        return None, None, None
-    elif len(all_species_names) == 0:
-        app_logger.error('No taxon found corresponding to %s', species)
+    nspecies = len(all_species_names)
+    if nspecies != 1:
+        app_logger.error('%s taxons found for %s', nspecies, species)
         return None, None, None
 
     return all_species_names[0]
 
 
 def get_species_from_sample(sample_name):
-    lims = _get_lims_connection()
-    samples = get_lims_samples(sample_name, lims)
-    species_string = None
+    samples = get_samples(sample_name)
     if samples:
         species_strings = set([s.udf.get('Species') for s in samples])
-        if len(species_strings) != 1:
-            app_logger.error('%s species found for sample %s', len(species_strings), sample_name)
-        else:
-            species_string = species_strings.pop()
-    if species_string:
-        taxid, scientific_name, common_name = get_species_information_from_ncbi(species_string)
-        if taxid:
-            return scientific_name
-    return None
+        nspecies = len(species_strings)
+        if nspecies != 1:
+            app_logger.error('%s species found for sample %s', nspecies, sample_name)
+            return None
+        species_string = species_strings.pop()
+        if species_string:
+            taxid, scientific_name, common_name = get_species_information_from_ncbi(species_string)
+            if taxid:
+                return scientific_name
 
 
 def sanitize_user_id(user_id):
     if isinstance(user_id, str):
         return re.sub("[^\w_\-.]", "_", user_id)
-    else:
-        return None
 
 
-def get_lims_samples(sample_name, lims):
+substitutions = (
+    (None, None),
+    (re.compile('_(\d{2})$'), ':\g<1>'),
+    (re.compile('__(\w):(\d{2})'), ' _\g<1>:\g<2>')
+)
+
+
+def get_list_of_samples(sample_names, sub=0):
+    pattern, repl = substitutions[sub]
+    _sample_names = list(sample_names)
+    if pattern and repl:
+        _sample_names = [pattern.sub(repl, s) for s in _sample_names]
+
+    lims = connection()
+    samples = lims.get_samples(name=_sample_names)
+    lims.get_batch(samples)
+
+    if len(samples) != len(sample_names):  # haven't got all the samples because some had _01/__L:01
+        if sub < len(substitutions):
+            remainder = sorted(set(_sample_names).difference(set([s.name for s in samples])))
+            samples.extend(get_list_of_samples(remainder, sub + 1))
+        else:
+            raise LimsCommunicationError('Expected %s back, got %s' % (_sample_names, len(samples)))
```

Comment:
No That I see this, I realise that this is going to fail (and crash) in case we have samples with the same names.
Is this what we want ?